### PR TITLE
Weight decay

### PR DESCRIPTION
**Pull request by amunozo**

### CODE REVIEW

File: train/trainer.py [hunk 1]
```diff
@@ -33,6 +34,8 @@ def __init__(self, model, train_params, start_time=None,
         self.last_checkpoint_i = -1
         self.last_checkpoint_nsamples = -1
         self.stat_syncer = 0
+        self.expected_batches_per_epoch = expected_batches_per_epoch
```
Comment (gailweiss) [inline train/trainer.py:37]: shouldnt this be in the linear scheduler commit?

File: train/trainer.py [hunk 2]
```diff
@@ -254,27 +260,58 @@ def reconfigure_optimizers(self):
...
-        def f_warmup(n):
-            assert n <= self.train_params.lr_warm_steps
-            return n / self.train_params.lr_warm_steps
+        if weight_decay > 0:
+            optimizer_grouped_parameters = self.get_optimizer_params(weight_decay)
+            optimizer = AdamW(optimizer_grouped_parameters, lr=self.train_params.lr)
+        else:
+            optimizer = torch.optim.Adam(self.parameters(), lr=self.train_params.lr)
 
+        def f_warmup(n):
```
Comment (gailweiss) [inline train/trainer.py:288]: stylistic preference, but please undo the change to f_warmup (unless you have a particular reason for it?). i like code to fail loudly if something unexpected happens (e.g. if i call f_warmup after the expected warmup period)

File: train/trainer.py [hunk 3]
```diff
@@ -254,27 +260,58 @@ def reconfigure_optimizers(self):
         optimizers, _ = self.configure_optimizers(
             existing_scheduler=self.lr_schedulers())
         self.optimizers()._optimizer = optimizers[0]
-
+    
+    def get_optimizer_params(self, weight_decay):
```
Comment (gailweiss) [inline train/trainer.py:264]: make this function work correctly even when weight_decay = 0, by adding an if at the beginning eg
```
if weight_decay <= 0:
    return self.parameters()
````

File: train/trainer.py [hunk 4]
```diff
@@ -254,27 +260,58 @@ def reconfigure_optimizers(self):
...
+    
     def configure_optimizers(self, existing_scheduler=None):
-        optimizer = torch.optim.Adam(self.parameters(),
-                                     lr=self.train_params.lr)
+        weight_decay = self.train_params.weight_decay
 
-        def f_warmup(n):
-            assert n <= self.train_params.lr_warm_steps
-            return n / self.train_params.lr_warm_steps
+        if weight_decay > 0:
```
Comment (gailweiss) [inline train/trainer.py:282]: once get_optimizer_params is updated to work even for weight decay zero, can make this code neater:
```
optimizer_params = self.get_optimizer_params(weight_decay)
optimizerClass = AdamW if weight_decay > 0 else Adam
optimizer = optimizerClass(optimizer_params, lr=self.train_params.lr)
```

File: train/trainer.py [hunk 5]
```diff
@@ -254,27 +260,58 @@ def reconfigure_optimizers(self):
...
+            if self.train_params.lr_warm_steps > 0:
+                return min(n / self.train_params.lr_warm_steps, 1.0)
+            else:
+                return 1.0
+        
         s_warmup = torch.optim.lr_scheduler.LambdaLR(optimizer, f_warmup)
         s_main = self.make_main_scheduler(optimizer)
-        s_full = MyChainedScheduler() if None is existing_scheduler else \
-            existing_scheduler
+        s_full = MyChainedScheduler() if existing_scheduler is None else existing_scheduler
```
Comment (gailweiss) [inline train/trainer.py:296]: goes past 80 characters

File: train/trainer.py [hunk 6]
```diff
@@ -254,27 +260,58 @@ def reconfigure_optimizers(self):
...
         s_full.setup(optimizer, [s_warmup, s_main],
-                     milestones=[self.train_params.lr_warm_steps])
-        # get scheduler started, else first batch has max value apparently
```
Comment (gailweiss) [inline train/trainer.py:272]: move this comment to just before line 299 ( `s_full.step(None)`)

File: train/trainer.py [hunk 7]
```diff
@@ -200,11 +200,14 @@ def training_step(self, batch, batch_idx):
         self.log("train_batch_loss", losses["main"].item())
         # for the lr scheduler
 
+        self.weight_norms = self.get_weight_norms()
```
Comment (gailweiss) [inline train/trainer.py:203]: why are you storing these instead of just running `self.log_stat("weight_norms", self.get_weight_norms())` below? are you using them somewhere?

File: train/trainer.py [hunk 8]
```diff
@@ -254,27 +257,59 @@ def reconfigure_optimizers(self):
...
+        optimizer_params = self.get_optimizer_params(weight_decay)
+        optimizerClass = torch.optim.AdamW if weight_decay > 0 else \
+            torch.optim.Adam
+        optimizer = optimizerClass(optimizer_params, lr=self.train_params.lr)
 
         def f_warmup(n):
             assert n <= self.train_params.lr_warm_steps
             return n / self.train_params.lr_warm_steps
-
+
```
Comment (gailweiss) [inline train/trainer.py:289]: pep8 - empty lines shouldnt have whitespace

File: train/trainer.py [hunk 9]
```diff
@@ -254,27 +257,59 @@ def reconfigure_optimizers(self):
...
-
+        
         s_warmup = torch.optim.lr_scheduler.LambdaLR(optimizer, f_warmup)
         s_main = self.make_main_scheduler(optimizer)
-        s_full = MyChainedScheduler() if None is existing_scheduler else \
+        s_full = MyChainedScheduler() if existing_scheduler is None else \
             existing_scheduler
         s_full.setup(optimizer, [s_warmup, s_main],
-                     milestones=[self.train_params.lr_warm_steps])
+                    milestones=[self.train_params.lr_warm_steps])
```
Comment (gailweiss) [inline train/trainer.py:295]: undo anti-pep8 style change
